src/realm/object-store/sync/app.cpp

```diff
@@ -862,106 +964,160 @@ std::string App::get_app_route(const Optional<std::string>& hostname) const
+                // retry_count is passed in to track the number of subsequent redirection attempts
+                self->request_location(std::move(completion), std::move(base_url), std::move(redir_location),
+                                       request.redirect_count);
+                return; // early return
             }
-            {
-                std::lock_guard<std::mutex> lock(*self->m_route_mutex);
-                self->m_location_updated = true;
+            // Location request was successful - update the location info
+            auto update_response = self->update_location(response, base_url);
+            if (update_response) {
+                self->log_error("App: request location failed (%1%2) - keeping original location information",
+                                update_response->code_string(),
+                                update_response->additional_status_code
+                                    ? util::format(" %1", *update_response->additional_status_code)
+                                    : "");
             }
-        }
-        catch (const AppError&) {
-            // Pass the response back to completion
-            return completion(response);
-        }
-        completion(util::none);
-    });
+            completion(update_response);
+        });
 }
 
-void App::post(std::string&& route, UniqueFunction<void(Optional<AppError>)>&& completion, const BsonDocument& body)
+Optional<AppError> App::update_location(const Response& response, const std::string& base_url)
 {
-    do_request(Request{HttpMethod::post, std::move(route), m_request_timeout_ms, get_request_headers(),
-                       Bson(body).to_string()},
-               handle_default_response(std::move(completion)));
+    // Check for errors in the response
+    if (auto error = AppUtils::check_for_errors(response)) {
+        return error;
+    }
+
+    REALM_ASSERT(m_sync_manager); // Need a valid sync manager
+
+    // Update the location info with the data from the response
+    try {
+        auto json = parse<BsonDocument>(response.body);
+        auto hostname = get<std::string>(json, "hostname");
+        auto ws_hostname = get<std::string>(json, "ws_hostname");
+        auto deployment_model = get<std::string>(json, "deployment_model");
+        auto location = get<std::string>(json, "location");
+        m_sync_manager->perform_metadata_update([&](SyncMetadataManager& manager) {
+            manager.set_app_metadata(deployment_model, location, hostname, ws_hostname, base_url);
+        });
+        // Update the local hostname and path information
+        update_hostname(hostname, ws_hostname, base_url);
+        {
+            std::lock_guard<std::mutex> lock(*m_route_mutex);
+            m_location_updated = true;
+            m_sync_manager->set_sync_route(make_sync_route());
+        }
+    }
+    catch (const AppError& ex) {
+        return ex;
```

Comment:
Also, handling these kinds of error chains is like what Futures/Promises were meant to do. Since this PR is re-writing a whole bunch of logic around calling out to the server to get info, you could consider making a futurized version of send_request_to_server that returns a Future<Response> and then chain the responses for each step of handling various kinds of responses and the promise/futures would take care of passing errors up the chain properly at each step. Maybe something like this rough structure?

```
send_request_to_server(location_request).then([this](Response resp) {
  // handle the app getting destroyed here
  if (resp.http_status_code == 300) {
    // handle redirect
    return send_request_to_server(new_redirected_location);
  }
  if (resp.http_status_code == 500) {
    return Status(ErrorCodes::OhNo, "Oh noes!");
  }
  // any other error handling here
  return resp;
})
.then( /* any other handlers you need to do while handling responses */)
.get_async([this](StatusWith<Response> resp) {
  if (resp.is_ok()) {
    update_location_metadata_with_response(resp);
  } else {
    handle_error(resp.get_status());
  }
});
```

Reply:
I agree - when I first started these changes, I went down the path of creating an `AppResponse` class that was similar to `StatusWith<Response>` and would automatically parse the response to check for an AppError (instead of doing it at the top of the chain).

One issue is that it will still need to be an AppError at the APP's API calls, otherwise it's a breaking change to all the SDKs.
Another issue is trying to pass the original `const Response& response` back up the chain without copying, esp. since some of the "function" AppServices calls could potentially return a lot of data, and since it's a const coming from the NetworkTransport's `send_request_to_server`, it can't be std::move()'d and removing the `const` is a breaking change to the SDKs (albeit rather minimal).

I'd prefer to make these changes in a separate PR for switching to a promise/future - if you agree, I can write a task for it and add it to my backlog, with a heads up to the SDKs that it is coming.